File: backend/app/services/task_repository.py
from __future__ import annotations

from datetime import datetime
from uuid import uuid4

from app.db.database import get_connection
from app.models import ExecutionRecord, Task


def _parse_datetime(value: str | None) -> datetime | None:
    return datetime.fromisoformat(value) if value else None


def _task_from_row(row) -> Task:
    return Task(
        id=row["id"],
        schedule=row["schedule"],
        session_id=row["session_id"],
        prompt=row["prompt"],
        enabled=bool(row["enabled"]),
        created_at=datetime.fromisoformat(row["created_at"]),
        updated_at=datetime.fromisoformat(row["updated_at"]),
        next_run_at=_parse_datetime(row["next_run_at"]),
    )


def _execution_from_row(row) -> ExecutionRecord:
    return ExecutionRecord(
        id=row["id"],
        task_id=row["task_id"],
        session_id=row["session_id"],
        prompt=row["prompt"],
        scheduled_for=datetime.fromisoformat(row["scheduled_for"]),
        executed_at=datetime.fromisoformat(row["executed_at"]),
        status=row["status"],
        error_message=row["error_message"],
    )


class TaskRepository:
    def list_tasks(self) -> list[Task]:
        with get_connection() as connection:
            rows = connection.execute(
                "SELECT * FROM tasks ORDER BY created_at DESC"
            ).fetchall()
        return [_task_from_row(row) for row in rows]

    def get_task(self, task_id: str) -> Task | None:
        with get_connection() as connection:
            row = connection.execute(
                "SELECT * FROM tasks WHERE id = ?",
                (task_id,),
            ).fetchone()
        return _task_from_row(row) if row else None

    def create_task(
        self,
        schedule: str,
        session_id: str,
        prompt: str,
        enabled: bool,
        created_at: datetime,
        next_run_at: datetime | None,
    ) -> Task:
        task_id = uuid4().hex
        with get_connection() as connection:
            connection.execute(
                """
                INSERT INTO tasks (
                    id, schedule, session_id, prompt, enabled, created_at, updated_at, next_run_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    task_id,
                    schedule,
                    session_id,
                    prompt,
                    int(enabled),
                    created_at.isoformat(),
                    created_at.isoformat(),
                    next_run_at.isoformat() if next_run_at else None,
                ),
            )
        return self.get_task(task_id)  # type: ignore[return-value]

    def update_task(
        self,
        task_id: str,
        schedule: str,
        prompt: str,
        updated_at: datetime,
        next_run_at: datetime | None,
    ) -> Task | None:
        with get_connection() as connection:
            connection.execute(
                """
                UPDATE tasks
                SET schedule = ?, prompt = ?, updated_at = ?, next_run_at = ?
                WHERE id = ?
                """,
                (
                    schedule,
                    prompt,
                    updated_at.isoformat(),
                    next_run_at.isoformat() if next_run_at else None,
                    task_id,
                ),
            )
        return self.get_task(task_id)

    def set_enabled(
        self,
        task_id: str,
        enabled: bool,
        updated_at: datetime,
        next_run_at: datetime | None,
    ) -> Task | None:
        with get_connection() as connection:
            connection.execute(
                """
                UPDATE tasks
                SET enabled = ?, updated_at = ?, next_run_at = ?
                WHERE id = ?
                """,
                (
                    int(enabled),
                    updated_at.isoformat(),
                    next_run_at.isoformat() if next_run_at else None,
                    task_id,
                ),
            )
        return self.get_task(task_id)

    def delete_task(self, task_id: str) -> None:
        with get_connection() as connection:
            connection.execute("DELETE FROM tasks WHERE id = ?", (task_id,))

    def list_due_tasks(self, now: datetime) -> list[Task]:
        with get_connection() as connection:
            rows = connection.execute(
                """
                SELECT * FROM tasks
                WHERE enabled = 1
                  AND next_run_at IS NOT NULL
                  AND next_run_at <= ?
                ORDER BY next_run_at ASC
                """,
                (now.isoformat(),),
            ).fetchall()
        return [_task_from_row(row) for row in rows]

    def update_next_run(self, task_id: str, next_run_at: datetime | None, updated_at: datetime) -> None:
        with get_connection() as connection:
            connection.execute(
                """
                UPDATE tasks
                SET next_run_at = ?, updated_at = ?
                WHERE id = ?
                """,
                (
                    next_run_at.isoformat() if next_run_at else None,
                    updated_at.isoformat(),
                    task_id,
                ),
            )

    def add_execution(
        self,
        task_id: str,
        session_id: str,
        prompt: str,
        scheduled_for: datetime,
        executed_at: datetime,
        status: str,
        error_message: str | None,
    ) -> None:
        with get_connection() as connection:
            connection.execute(
                """
                INSERT INTO execution_history (
                    task_id, session_id, prompt, scheduled_for, executed_at, status, error_message
                ) VALUES (?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    task_id,
                    session_id,
                    prompt,
                    scheduled_for.isoformat(),
                    executed_at.isoformat(),
                    status,
                    error_message,
                ),
            )

    def list_execution_history(self, task_id: str | None = None) -> list[ExecutionRecord]:
        query = "SELECT * FROM execution_history"
        params: tuple[str, ...] = ()
        if task_id:
            query += " WHERE task_id = ?"
            params = (task_id,)
        query += " ORDER BY executed_at DESC"
        with get_connection() as connection:
            rows = connection.execute(query, params).fetchall()
        return [_execution_from_row(row) for row in rows]

File: backend/src/task-repository.ts
import type Database from "better-sqlite3";

import type { ExecutionRecord, ExecutionStatus, Task } from "./types.js";

function parseEnvironmentVariables(value: unknown): Record<string, string> {
  if (typeof value !== "string" || !value.trim()) {
    return {};
  }

  try {
    const parsed = JSON.parse(value) as unknown;
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      return {};
    }

    return Object.fromEntries(
      Object.entries(parsed).filter((entry): entry is [string, string] => typeof entry[0] === "string" && typeof entry[1] === "string")
    );
  } catch {
    return {};
  }
}

function parseTaskRow(row: Record<string, unknown> | undefined): Task | null {
  if (!row) {
    return null;
  }

  return {
    id: String(row.id),
    schedule: String(row.schedule),
    thread_id: String(row.thread_id),
    prompt: String(row.prompt),
    workspace_directory: String(row.workspace_directory),
    environment_variables: parseEnvironmentVariables(row.environment_variables),
    enabled: Boolean(row.enabled),
    created_at: String(row.created_at),
    updated_at: String(row.updated_at),
    next_run_at: row.next_run_at ? String(row.next_run_at) : null
  };
}

function parseExecutionRow(row: Record<string, unknown>): ExecutionRecord {
  return {
    id: Number(row.id),
    task_id: String(row.task_id),
    thread_id: String(row.thread_id),
    prompt: String(row.prompt),
    scheduled_for: String(row.scheduled_for),
    executed_at: String(row.executed_at),
    status: String(row.status) as ExecutionStatus,
    error_message: row.error_message ? String(row.error_message) : null
  };
}

type CreateTaskParams = {
  id: string;
  schedule: string;
  threadId: string;
  prompt: string;
  workspaceDirectory: string;
  environmentVariables: Record<string, string>;
  enabled: boolean;
  createdAt: string;
  nextRunAt: string | null;
};

type UpdateTaskParams = {
  schedule: string;
  prompt: string;
  environmentVariables: Record<string, string>;
  updatedAt: string;
  nextRunAt: string | null;
};

type AddExecutionParams = {
  taskId: string;
  threadId: string;
  prompt: string;
  scheduledFor: string;
  executedAt: string;
  status: ExecutionStatus;
  errorMessage: string | null;
};

export class TaskRepository {
  database: Database.Database;

  constructor(database: Database.Database) {
    this.database = database;
  }

  listTasks(): Task[] {
    const rows = this.database
      .prepare("SELECT * FROM tasks ORDER BY created_at DESC")
      .all() as Array<Record<string, unknown>>;
    return rows.map((row) => parseTaskRow(row)).filter((task): task is Task => task !== null);
  }

  getTask(taskId: string): Task | null {
    const row = this.database
      .prepare("SELECT * FROM tasks WHERE id = ?")
      .get(taskId) as Record<string, unknown> | undefined;
    return parseTaskRow(row);
  }

  createTask({ id, schedule, threadId, prompt, workspaceDirectory, environmentVariables, enabled, createdAt, nextRunAt }: CreateTaskParams): Task {
    this.database
      .prepare(`
        INSERT INTO tasks (
          id, schedule, thread_id, prompt, workspace_directory, environment_variables, enabled, created_at, updated_at, next_run_at
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
      `)
      .run(
        id,
        schedule,
        threadId,
        prompt,
        workspaceDirectory,
        JSON.stringify(environmentVariables),
        enabled ? 1 : 0,
        createdAt,
        createdAt,
        nextRunAt
      );
    return this.getTask(id) as Task;
  }

  updateTask(taskId: string, { schedule, prompt, environmentVariables, updatedAt, nextRunAt }: UpdateTaskParams): Task | null {
    this.database
      .prepare(`
        UPDATE tasks
        SET schedule = ?, prompt = ?, environment_variables = ?, updated_at = ?, next_run_at = ?
        WHERE id = ?
      `)
      .run(schedule, prompt, JSON.stringify(environmentVariables), updatedAt, nextRunAt, taskId);
    return this.getTask(taskId);
  }

  setEnabled(taskId: string, enabled: boolean, updatedAt: string, nextRunAt: string | null): Task | null {
    this.database
      .prepare(`
        UPDATE tasks
        SET enabled = ?, updated_at = ?, next_run_at = ?
        WHERE id = ?
      `)
      .run(enabled ? 1 : 0, updatedAt, nextRunAt, taskId);
    return this.getTask(taskId);
  }

  deleteTask(taskId: string): void {
    this.database.prepare("DELETE FROM tasks WHERE id = ?").run(taskId);
  }

  listDueTasks(nowIso: string): Task[] {
    const rows = this.database
      .prepare(`
        SELECT * FROM tasks
        WHERE enabled = 1
          AND next_run_at IS NOT NULL
          AND next_run_at <= ?
        ORDER BY next_run_at ASC
      `)
      .all(nowIso) as Array<Record<string, unknown>>;
    return rows.map((row) => parseTaskRow(row)).filter((task): task is Task => task !== null);
  }

  updateNextRun(taskId: string, nextRunAt: string | null, updatedAt: string): void {
    this.database
      .prepare(`
        UPDATE tasks
        SET next_run_at = ?, updated_at = ?
        WHERE id = ?
      `)
      .run(nextRunAt, updatedAt, taskId);
  }

  addExecution({ taskId, threadId, prompt, scheduledFor, executedAt, status, errorMessage }: AddExecutionParams): void {
    this.database
      .prepare(`
        INSERT INTO execution_history (
          task_id, thread_id, prompt, scheduled_for, executed_at, status, error_message
        ) VALUES (?, ?, ?, ?, ?, ?, ?)
      `)
      .run(taskId, threadId, prompt, scheduledFor, executedAt, status, errorMessage);
  }

  listExecutionHistory(taskId: string | null = null): ExecutionRecord[] {
    if (taskId) {
      const rows = this.database
        .prepare("SELECT * FROM execution_history WHERE task_id = ? ORDER BY executed_at DESC")
        .all(taskId) as Array<Record<string, unknown>>;
      return rows.map(parseExecutionRow);
    }

    const rows = this.database
      .prepare("SELECT * FROM execution_history ORDER BY executed_at DESC")
      .all() as Array<Record<string, unknown>>;
    return rows.map(parseExecutionRow);
  }
}
